File: 프로그래머스/lv2/87946. 피로도/피로도.swift
import Foundation

func solution(_ k:Int, _ dungeons:[[Int]]) -> Int {

    var answer = 0
    var visited = Array(repeating: false, count: dungeons.count)

    func dfs(_ count: Int, _ k: Int) {
        if count > dungeons.count {
            return
        }
        answer = max(answer, count)

        for i in 0..<dungeons.count {
            if !visited[i] && k >= dungeons[i][0] {
                visited[i] = true
                dfs(count+1, k - dungeons[i][1])
                visited[i] = false
            }
        }
    }

    dfs(0, k)

    return answer
}

File: 프로그래머스/2/87946. 피로도/피로도.py
answer = -1

def solution(k, dungeons):
    visited = [False] * len(dungeons)
    
    def backtrakcing(k, count):
        
        global answer
        answer = max(answer, count)
        
        for i in range(len(dungeons)):
            if not visited[i] and dungeons[i][0] <= k:
                visited[i] = True
                backtrakcing(k - dungeons[i][1], count+1)
                visited[i] = False
                
    backtrakcing(k, 0)
                
    return answer